### PR TITLE
First PoC iteration for Docker build

### PR DESCRIPTION
**Pull request by stp-ip**

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+RUN apk --no-cache add ca-certificates \
+  && apk --no-cache add \
+    curl \
+    tar
+
+RUN mkdir /source
+WORKDIR /source
+RUN curl -Lo strichliste.tar.gz https://github.com/strichliste/server/releases/download/v0.2.3/strichliste.tar.gz
+RUN tar -xf strichliste.tar.gz
+RUN rm -r strichliste.tar.gz
```
Comment (foosinn) [inline docker/Dockerfile]: I would prefer to put the Dockerfile in the project root and use `ADD . /source` and `.dockerignore`. 
Using commands to download files leads to issues with dockers caching. If you really want to download it you can use the `ADD <url>` command.

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    php7-pdo_mysql \
+    php7-fpm \
+    nginx
+
+COPY --from=release /source source
+
+COPY entrypoint.sh /source/entrypoint.sh
+RUN chmod +x /source/entrypoint.sh
+
+RUN adduser -u 82 -D -S -G www-data www-data
```
Comment (foosinn) [inline docker/Dockerfile]: Maybe we should use a environment variable to supply the `uid` and `gid`? That simplifies migrations and backups.

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+RUN cp /source/var/app.db.example /source/app.db.example
+RUN cp /source/var/app.db.example /source/var/app.db
+
+VOLUME /source/var
+
+WORKDIR /source/public
+EXPOSE 8080
+
+ENTRYPOINT ["/source/entrypoint.sh"]
+CMD nginx && php-fpm7
```
Comment (foosinn) [inline docker/Dockerfile]: Not sure if wanted here, but i like to split nignx and php-fpm to seperate containers and use docker-compose.

Comment (stp-ip) [reply docker/Dockerfile]: Not everyone has docker compose or uses that so as a "getting started" option the bundling is reasonable in my view.
More advanced setups can be done differently. It's all in the dockerfile so easy to throw out nginx.

File: docker/Dockerfile (hunk 2 above)
Comment (foosinn) [reply docker/Dockerfile]: I usually provide default values in the Dockerfile. If you create the user in the entrypoint script there should not be any pitfalls.

```
ENV UID=1000 \
    GID=1000
```

File: docker/Dockerfile (hunk 1 above)
Comment (foosinn) [reply docker/Dockerfile]: AFAIK `docker build` will use a cache if no change was made to the Dockerfile and the upstream image was not pulled.

File: docker/Dockerfile (hunk 2 above)
Comment (stp-ip) [reply docker/Dockerfile]: Both `useradd` and `chown` need root. So using `USER www-data` wouldn't be possible and the container starts and runs as root. That's the primary argument against setting them as ENVs in my view.

File: docker/Dockerfile (hunk 1 above)
Comment (stp-ip) [reply docker/Dockerfile]: As we are using versioned releases, which should be immutable, a new release means a change to the dockerfile and therefore clearing the cache. But I might be missing something.

**Comment by schinken** [comment]
Actually, I would prefer that this docker image:
* Is "production ready"
* Is using docker compose with a separate nginx/fpm container

Also we can think about using a "real database" instead of sqlite, if that helps with the uid/gid mapping

**Comment by stp-ip** [comment]
> Actually, I would prefer that this docker image:
> 
>     * Is "production ready"

In my view the docker image is "production ready".
We can definitely simplify it by running as root, that removes the uid and gid burden partly, but in my view that's less "production ready".

>     * Is using docker compose with a separate nginx/fpm container

Single container vs separate nginx and fpm container needs additional tooling and understanding of docker compose, also prevents it running on schedulers such as managed providers such as ECS or Kubernetes (GKE, AKE).

> Also we can think about using a "real database" instead of sqlite, if that helps with the uid/gid mapping

The burden is shifted to the database. It's either another container (still needs host permissions) or manually done on the host. Either way a user needs to be created (even just by the package manager), which in my view makes it similar to just letting the user create a www-data user, which in most distributions is one that is available already.

**Comment by schinken** [comment]
I have to close that pullrequest because I force-pushed to that repo because I removed a file from the history. now the pullrequest is fucked up. sorry for that .(

**Pull request closed without merging**